### PR TITLE
[transfer-to-object] TTO support in test scenario

### PR DESCRIPTION
**Pull request by tzakian**
## Description 

This adds support for transfer-to-object/Receiving tickets in the test scenario.

In order to do this a "proper" object store needed to be added to the test scenario since we cannot count on caching to keep the object for us like we can with dynamic fields in test scenario tests.

This adds such a test by creating a shared `InMemoryStorage` (wrapped in a mutex since tests can be multi-threaded and we need internal mutability). Two handles are then created for this -- one for the storage read operations, and another is added to the context extension so that we can place "objects-to-be-received" into this store when we create the `Receiving` ticket.

This might generally be useful for other tests as well/refactoring the test scenario to be more realistic but I will leave any such investigations for another day.

## Test Plan 

Added tests for the new features to the test scenario tests.

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

**Comment by tnowacki** [comment]
Am I losing my mind or did I review this already? Or something similar?

File: crates/sui-framework/packages/sui-framework/sources/test/test_scenario.move
Comment (tnowacki) [inline crates/sui-framework/packages/sui-framework/sources/test/test_scenario.move:368]: Is this just for receiving? I feel like this will be very easily conflated with dynamic fields

File: crates/sui-framework/packages/sui-framework/sources/test/test_scenario.move
Comment (tnowacki) [inline crates/sui-framework/packages/sui-framework/sources/test/test_scenario.move:376]: Is this the right error code?

File: crates/sui-framework/packages/sui-framework/sources/test/test_scenario.move
Comment (tnowacki) [inline crates/sui-framework/packages/sui-framework/sources/test/test_scenario.move:372]: I find this name a bit confusing, since I didn't immediately tell it was for the parent?

Comment (tnowacki) [reply crates/sui-framework/packages/sui-framework/sources/test/test_scenario.move:372]: It doesn't seem to line up with `get_receiving_ticket_for_object_by_id`

**Comment by tzakian** [comment]
> Am I losing my mind or did I review this already? Or something similar?
I don't think so? But maybe I'm losing my mind too...

**Comment by amnn** [comment]
> Am I losing my mind or did I review this already? Or something similar?

Shared object deletion support probably?

File: crates/sui-framework/packages/sui-framework/sources/test/test_scenario.move
Comment (amnn) [inline crates/sui-framework/packages/sui-framework/sources/test/test_scenario.move:399]: ditto to Todd's comment about error codes here.

File: crates/sui-framework/packages/sui-framework/tests/test_scenario_tests.move
Comment (amnn) [inline crates/sui-framework/packages/sui-framework/tests/test_scenario_tests.move:532]: These tests could maybe benefit from a short comment explaining what they are testing -- I see there's an effort made to have the `name` do that job, but it's stretching the limits of what is feasible with a Move identifier today...

File: sui-execution/latest/sui-move-natives/src/test_scenario.rs
Comment (amnn) [inline sui-execution/latest/sui-move-natives/src/test_scenario.rs:118]: Does this comment need updating, or is it related to the other empty map?

File: sui-execution/latest/sui-move-natives/src/test_scenario.rs
Comment (amnn) [inline sui-execution/latest/sui-move-natives/src/test_scenario.rs:587]: Given the native functions return these error codes, I'm now a little confused why the Move code returns error codes as well -- do we need both?

File: sui-execution/latest/sui-move-natives/src/test_scenario.rs
Comment (amnn) [inline sui-execution/latest/sui-move-natives/src/test_scenario.rs:607]: I'm kind of surprised that you need to create this `MoveObject`/`Object` here -- I would expect the object to exist already, so that you can take it from there and wrap it in the `Receiving<T>`, is that not the case?

File: external-crates/move/crates/move-vm-runtime/src/native_extensions.rs
Comment (amnn) [inline external-crates/move/crates/move-vm-runtime/src/native_extensions.rs:1]: Is this change here to stay? Seems like debug code and I can't spot a caller in the PR.

Comment (tzakian) [reply external-crates/move/crates/move-vm-runtime/src/native_extensions.rs:1]: Nope, some debug code that slipped past. Will remove.

File: sui-execution/latest/sui-move-natives/src/test_scenario.rs
Comment (tzakian) [reply sui-execution/latest/sui-move-natives/src/test_scenario.rs:118]: Will update!

File: sui-execution/latest/sui-move-natives/src/test_scenario.rs
Comment (tzakian) [reply sui-execution/latest/sui-move-natives/src/test_scenario.rs:607]: The joys of test scenario...

It is sadly not the case -- the test inventories store Move values, which makes sense in every case except this one generally.

Eventually I think we should rewrite test scenario to use an actual storage layer, and simply provide hooks into that, but at least until then...

File: sui-execution/latest/sui-move-natives/src/test_scenario.rs
Comment (tzakian) [reply sui-execution/latest/sui-move-natives/src/test_scenario.rs:587]: Nope! Will remove `was_allocated_receiving_ticket` fn :)

File: crates/sui-framework/packages/sui-framework/tests/test_scenario_tests.move
Comment (tzakian) [reply crates/sui-framework/packages/sui-framework/tests/test_scenario_tests.move:532]: It's like a tweet, how much can you fit into 128 characters...

**Comment by tzakian** [comment]
@tnowacki and @amnn just bumping this to see if we can get this in soon/if there's anything left outstanding before we can land this. Have been fielding some requests for adding TTO support in test scenario.

File: crates/sui-framework/packages/sui-framework/tests/test_scenario_tests.move
Comment (amnn) [inline crates/sui-framework/packages/sui-framework/tests/test_scenario_tests.move:413]: Move 2024 time for these now?

File: crates/sui-framework/packages/sui-framework/tests/test_scenario_tests.move
Comment (amnn) [inline crates/sui-framework/packages/sui-framework/tests/test_scenario_tests.move:449]:
```suggestion
    // Happy path test: Receive a single object from an object in a transaction.
```

File: crates/sui-framework/packages/sui-framework/sources/test/test_scenario.move
Comment (amnn) [inline crates/sui-framework/packages/sui-framework/sources/test/test_scenario.move:369]: Is there potential for confusion by using the child/parent nomenclature to refer to objects sent to other objects, seeing as it is also used for the dynamic field internals? Not entirely sure what the better option is here (... `owner` and `receivable`?)

**Review by amnn — COMMENTED** [review]
My only comment is an annoying bikeshedding comment about naming things, but I think @tnowacki had more useful thoughts/comments on the actual implementation, so I'll defer to him for the final accept -- it otherwise looks good to me though!

File: crates/sui-framework/packages/sui-framework/sources/test/test_scenario.move
Comment (tzakian) [reply crates/sui-framework/packages/sui-framework/sources/test/test_scenario.move:369]: More than happy to bikeshed on the nomenclature here -- I've always referred to them as/the best I've been able to come up with is: `parent`/`child`, or `object`/`object-owned-object` (although I tend to not use this last one since it's a bit long...). 

`owner`/`receivable` might be nice since with the addition of private receive rules for `key`-only objects, you don't have "ownership" over the object in the sense that you're guaranteed access to it. + `receivable_ids_for_owner_id` has a nice ring to it.

Comment (tnowacki) [reply crates/sui-framework/packages/sui-framework/sources/test/test_scenario.move:369]: +1 to no parent/child nomenclature. It is really an implementation detail from the Move perspective.

File: crates/sui-framework/packages/sui-framework/sources/test/test_scenario.move
Comment (tnowacki) [inline crates/sui-framework/packages/sui-framework/sources/test/test_scenario.move:405]: nit, doc comments

File: crates/sui-framework/packages/sui-framework/sources/test/test_scenario.move
Comment (tnowacki) [inline crates/sui-framework/packages/sui-framework/sources/test/test_scenario.move:36]: soon™️

File: sui-execution/latest/sui-move-natives/src/test_scenario.rs
Comment (tnowacki) [inline sui-execution/latest/sui-move-natives/src/test_scenario.rs:54]: I don't quite understand the Mutex here. Can you link to the relevant code? (what goes wrong if it is missing)

File: sui-execution/latest/sui-move-natives/src/test_scenario.rs
Comment (tzakian) [inline sui-execution/latest/sui-move-natives/src/test_scenario.rs:632]: @tnowacki This is where the mutex is needed. Since we need to grab mutable access to the inner `InMemoryStorage`.

File: sui-execution/latest/sui-move-natives/src/test_scenario.rs
Comment (tzakian) [reply sui-execution/latest/sui-move-natives/src/test_scenario.rs:54]: This is needed because of how this is used. 

In `unit_test.rs` one of these is constructed and placed in `TEST_STORE`. 

The `Mutex` is allowing us to then access the inner `InMemoryStorage` here to place objects to then be received. I'll tag the places where this is used.

File: crates/sui-move/src/unit_test.rs
Comment (tzakian) [inline crates/sui-move/src/unit_test.rs:153]: Adding it to the extensions so that we can access it to add objects to it later when we allocate a receiving ticket.

File: sui-execution/latest/sui-move-natives/src/test_scenario.rs
Comment (tnowacki) [reply sui-execution/latest/sui-move-natives/src/test_scenario.rs:632]: Why do you need a mutex? versus something like a `RefCell`?

Comment (tnowacki) [reply sui-execution/latest/sui-move-natives/src/test_scenario.rs:632]: And can you not use `context.extensions_mut` here?

File: crates/sui-framework/packages/sui-framework/sources/test/test_scenario.move
Comment (tnowacki) [inline crates/sui-framework/packages/sui-framework/sources/test/test_scenario.move:372]:
```suggestion
    public fun most_recent_receiving_ticket<T: key>(
```
Feels more in line with the functions above?

File: crates/sui-framework/packages/sui-framework/sources/test/test_scenario.move
Comment (tnowacki) [inline crates/sui-framework/packages/sui-framework/sources/test/test_scenario.move:366]:
```suggestion
    public fun receivable_objects_for_owner_id<T: key>(object: ID): vector<ID> {
```
Just trying to make some of these easier to parse

File: crates/sui-framework/packages/sui-framework/sources/test/test_scenario.move
Comment (tnowacki) [inline crates/sui-framework/packages/sui-framework/sources/test/test_scenario.move:383]:
```suggestion
    public fun receiving_ticket_by_id<T: key>(
```

**Review by tnowacki — APPROVED** [review]
LGTM. Just trying to shorten these names up a bit. The amount of words and repeated use of "object" in some of them makes them a bit strange to parse mentally

**Pull request merged**
